### PR TITLE
resolved: strict mode for DNS-over-TLS

### LINKED ISSUE #10755: systemd-resolved: "strict" resolution for DNSOverTLS?
It seems systemd-resolved currently only supports "opportunistic" DNSOverTLS resolution. From a privacy point of view, if I enable DNSOverTLS I'd most likely expect it to work over TLS, or fail (otherwise one would be subject to downgrade attacks).

Is there any plan to support "strict" mode, other than "opportunistic"?

Thanks,
Fernando

### PR DESCRIPTION
**Pull request by irtimmer**
Better late than never. I've finally finished the main parts to support certificate checking for DNS-over-TLS. This fixes #10755 and partially #9397. This is the minimum amount of changes to support this feature and some other small changes, like support for TLS 1.3 with GnuTLS and moving some fields from DnsServer to Manager. Therefore it will only work with server certificates signed for the server IP-address and the list of trusted root certificates can't be configured.

_p.s. I hope to send soon a separate PR to support certificate checking with domain name or  SPKI hash._

### CODE REVIEW

File: src/resolve/resolved-manager.h [hunk 1]
```diff
@@ -18,6 +18,7 @@ typedef struct Manager Manager;
 #include "resolved-dns-server.h"
 #include "resolved-dns-stream.h"
 #include "resolved-dns-trust-anchor.h"
+#include "resolved-dnstls.h"
```
Comment (yuwata) [inline src/resolve/resolved-manager.h:21]: Please add `#if ENABLE_DNS_OVER_TLS`.

BTW, I think the `#if` condition should be moved to `resolved-dnstls.h`.

File: src/resolve/resolved-dnstls-openssl.c [hunk 2]
```diff
@@ -51,6 +51,8 @@ static int dnstls_flush_write_buffer(DnsStream *stream) {
 int dnstls_stream_connect_tls(DnsStream *stream, DnsServer *server) {
         _cleanup_(BIO_freep) BIO *rb = NULL, *wb = NULL;
         _cleanup_(SSL_freep) SSL *s = NULL;
+        X509_VERIFY_PARAM *v;
+        const unsigned char *ip;
```
Comment (poettering) [inline src/resolve/resolved-dnstls-openssl.c]: doesn't matter much, but please move into the if block where this is needed. we try to keep most variables declared at minimal scope

File: src/resolve/resolved-manager.h (hunk 1 above)
Comment (irtimmer) [reply src/resolve/resolved-manager.h:21]: I've moved the condition with commit 0b5c09b

**Comment by irtimmer** [comment]
I've made some small changes. Now only the LGTM analysis fails because it's unable to fetch the new commits for some unclear reason.

**Comment by evverx** [comment]
@irtimmer it looks like it was glitch. As far as I can tell, it had something to do with authentication via GitHub. Everything should be fine now.

File: src/resolve/resolved-dns-server.h [hunk 3]
```diff
@@ -37,9 +37,7 @@ int dns_server_feature_level_from_string(const char *s) _pure_;
 
 #include "resolved-link.h"
 #include "resolved-manager.h"
-#if ENABLE_DNS_OVER_TLS
 #include "resolved-dnstls.h"
```
Comment (yuwata) [inline src/resolve/resolved-dns-server.h]: Then, please sort the headers.

File: src/resolve/resolved-dns-stream.h [hunk 4]
```diff
@@ -17,9 +17,7 @@ typedef enum DnsStreamType {
 #include "resolved-dns-packet.h"
 #include "resolved-dns-transaction.h"
 #include "resolved-manager.h"
-#if ENABLE_DNS_OVER_TLS
 #include "resolved-dnstls.h"
-#endif
```
Comment (yuwata) [inline src/resolve/resolved-dns-stream.h:22]: Also here.

File: src/resolve/resolved-dns-transaction.c [hunk 5]
```diff
@@ -12,9 +12,7 @@
 #include "resolved-dns-cache.h"
 #include "resolved-dns-transaction.h"
 #include "resolved-llmnr.h"
-#if ENABLE_DNS_OVER_TLS
 #include "resolved-dnstls.h"
-#endif
```
Comment (yuwata) [inline src/resolve/resolved-dns-transaction.c:17]: Also here.

**Comment by poettering** [comment]
looks dope, besides the include ordering thing.

**Pull request merged**